opal/mca/threads/wait_sync.h

```diff
@@ -22,13 +30,108 @@
 #define OPAL_MCA_THREADS_WAIT_SYNC_H
 
 #include "opal/mca/threads/condition.h"
+#include "opal/mca/threads/mutex.h"
+#include "opal/mca/threads/threads.h"
 #include "opal/sys/atomic.h"
 
 BEGIN_C_DECLS
 
 extern int opal_max_thread_in_progress;
 
-#include MCA_threads_wait_sync_base_include_HEADER
+typedef struct ompi_wait_sync_t {
+    opal_atomic_int32_t count;
+    int32_t status;
+    opal_thread_internal_cond_t condition;
+    opal_thread_internal_mutex_t lock;
+    struct ompi_wait_sync_t *next;
+    struct ompi_wait_sync_t *prev;
+    volatile bool signaling;
+} ompi_wait_sync_t;
+
+#define SYNC_WAIT(sync) (opal_using_threads() ? ompi_sync_wait_mt(sync) : sync_wait_st(sync))
+
+/* The loop in release handles a race condition between the signaling
+ * thread and the destruction of the condition variable. The signaling
+ * member will be set to false after the final signaling thread has
+ * finished operating on the sync object. This is done to avoid
+ * extra atomics in the signalling function and keep it as fast
+ * as possible. Note that the race window is small so spinning here
+ * is more optimal than sleeping since this macro is called in
+ * the critical path. */
+#define WAIT_SYNC_RELEASE(sync)                                  \
+    if (opal_using_threads()) {                                  \
+        while ((sync)->signaling) {                              \
+            if (OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT) {           \
+                opal_thread_yield();                             \
+            }                                                    \
+            continue;                                            \
+        }                                                        \
+        opal_thread_internal_cond_destroy(&((sync)->condition)); \
+        opal_thread_internal_mutex_destroy(&((sync)->lock));     \
```

Comment:
@shintaro-iwasaki  These additional parans that you added in your last commit really aren't needed. Also, we call this macro as a function (with a trailing semicolon) so they should have the canonical `do { ... } while(0)` form. Alternatively, maybe we should map them to a static inline function call. I don't see any reason for having them, everything has a known type here.

Reply:
I copied this from the original Pthreads code, so please be patient.

Anyway, I will do refactoring for these macros.  What is your suggestion?
1. Keep macros and use `do { ... } while(0)`.
2. Use `static inline` functions but keep the capitalized names (e.g., `static inline void WAIT_SYNC_RELEASE()'`).
3. Use `static inline` functions and rename all the functions (`WAIT_SYNC_RELEASE` becomes `opal_thread_wait_sync_release()`).

If we keep a macro form, what is your suggestion regarding the variable name:
 A. Keep the existing ones (the format is not unified, so I don't like it).
 B. `opal_thread_internal_cond_destroy(&((sync)->condition));`  (This is what I did.)
 C. `opal_thread_internal_cond_destroy(&sync->condition);`  (This can be incorrect if `sync` is expanded unexpectedly.)
 D. `opal_thread_internal_cond_destroy(&(sync->condition));`  (This can be incorrect if `sync` is expanded unexpectedly.)
 E. `opal_thread_internal_cond_destroy(&(sync)->condition);`  (@abouteiller pointed out this is confusing.)

Reply:
Is there any legitimate reason to change these macros to inlined functions ? If yes, then you should alter their names to only have lower case letter, only macros are in caps (OMPI standard coding practice).

Reply:
So should I
 1. keep the macro form (I will use `do { ... } while(0)`)

and unify the argument format to

  - E. `opal_thread_internal_cond_destroy(&(sync)->condition);`

since (as @devreal said)
> These additional parans that you added in your last commit really aren't needed.

?

If no one opposes, I will follow this.

Reply:
It seems that the entire discussion started from a typo in the code. If you put the closing parentheses at the right place (like in all the other uses around this code), everything will be good.

Reply:
I fixed the parentheses (please check the last commit). Please check them.